src/Halogen.purs

```diff
@@ -1,159 +1,78 @@
-
--- | A variant of `runUI` which supports a _post-render hook_. This allows applications
--- | to support third-party components or other custom behaviors by modifying the DOM after
--- | each update.
--- |
--- | This is considered an advanced feature, and should only be used with an understanding of
--- | the rendering pipeline.
-runUIWith :: forall req eff.
-               Component (Event (HalogenEffects eff)) req req ->
-               (req -> HTMLElement -> Driver req eff -> Eff (HalogenEffects eff) Unit) ->
-               Eff (HalogenEffects eff) (Tuple HTMLElement (Driver req eff))
-runUIWith sf postRender = mainLoop (pure <<< componentProcess sf postRender)
-
--- | A `Process` receives inputs and outputs effectful computations which update the DOM.
-type Process req eff = SF (Tuple req HTMLElement) (Eff (HalogenEffects eff) HTMLElement)
-
--- | Build a `Process` from a `Component`.
-componentProcess :: forall req eff.
-                      Component (Event (HalogenEffects eff)) req req ->
-                      (req -> HTMLElement -> Driver req eff -> Eff (HalogenEffects eff) Unit) ->
-                      Driver req eff ->
-                      Tuple HTMLElement (Process req eff)
-componentProcess sf postRender driver =
-  let render  = R.renderHTML requestHandler
-      vtrees  = render <$> sf
-      diffs   = tail vtrees >>> changes (head vtrees)
-      process = (diffs &&& input) *** input
-      node    = createElement (head vtrees)
-  in Tuple node (applyPatch <$> process)
-  where
-  requestHandler :: Event (HalogenEffects eff) req -> Eff (HalogenEffects eff) Unit
-  requestHandler aff = runEvent logger driver aff
-
-  logger :: Error -> Eff (HalogenEffects eff) Unit
-  logger e = log $ "Uncaught error in asynchronous code: " <> message e
-
-  applyPatch :: Tuple (Tuple Patch req) HTMLElement -> Eff (HalogenEffects eff) HTMLElement
-  applyPatch (Tuple (Tuple p req) node) = do
-    node' <- patch p node
-    postRender req node' driver
-    return node'
+import Control.Monad.Eff (Eff())
+import Control.Monad.Eff.Exception (throwException, error)
+import Control.Monad.Eff.Ref (Ref(), newRef, writeRef, readRef)
+import Control.Monad.Free
+import Control.Monad.State (runState)
+import Control.Monad.State.Trans (runStateT)
+
+import Data.Coyoneda (Coyoneda(), liftCoyoneda)
+import Data.DOM.Simple.Types (HTMLElement())
+import Data.Inject (Inject, inj)
+import Data.Maybe (Maybe(..))
+import Data.Tuple (Tuple(..))
+import Data.Void (Void())
+
+import Halogen.Component (Component(), renderComponent, queryComponent)
+import Halogen.Effects (HalogenEffects())
+import Halogen.HTML.Renderer.VirtualDOM (renderHTML)
+import Halogen.Internal.VirtualDOM (VTree(), createElement, diff, patch)
+
+type DriverState s =
+  Maybe { node :: HTMLElement
+        , vtree :: VTree
+        , state :: s
+        }
+
+type Driver f eff = forall i. f i -> Eff (HalogenEffects eff) i
+
+runUI :: forall eff s f. Component s f (Eff (HalogenEffects eff)) Void
```

Comment:
Ah right, yes, I wanted to talk about this (`Aff` vs `Eff`, I'll reply to MonadAff separately)

One slightly unfortunate thing about using `Aff` over `Eff` is since Halogen then has to run the `Aff`s it means that if there's a failure and `attempt` wasn't used the error kinda disappears (in the current Halogen we just log it to the console).

If you use `Eff` instead, it means one has to `runAff` themselves, and there's no way that error can disappear, unless `launchAff` was used, which is the expected behaviour anyway.

I'm totally fine with using `Aff`, just wanted to discuss this tradeoff and forgot to mention it in the PR.